### ISSUE TITLE
keyapi

### ISSUE THREAD

**Issue by arekinath** [joyent/sdc-cloudapi#1]
After my sdc-discuss post, @marsell followed up and said I should file an issue about this, so here it is:

Currently in `lib/auth.js` there is an implementation of Token-based auth for CloudAPI using `keyapi` and `sdc-securetoken`. `app.js` inserts the function handling this step, `tokenAuth`, into the chain after `signatureAuth`, and at present there's no way to escape from `signatureAuth` that I can see without an error. So in the code as it is I can't find any way to actually use this token auth mechanism to authenticate a request to CloudAPI.

Adding two lines at the top of `signatureAuth` changes the story however:

``` javascript
if (req.header('X-Auth-Token'))
  return next();
```

Then I can successfully use `sdc-securetoken` to generate some JSON to put in `X-Auth-Token` and authenticate with the current CloudAPI.

Now I'm given to understand that this code is a legacy left over from previous JPC development, which is why it's undocumented and the `keyapi` repo is still private. However, I'm sure I'm not the only one with a private SDC deployment who is interested in developing their own "customer portal".

At present, the only SDC API that is guaranteed to stay stable over time is CloudAPI, so if developing a customer portal outside of Joyent, sticking to CloudAPI as much as possible seems like the most sensible approach. 

However, with the only documented form of authentication against CloudAPI requiring the use of RSA keys, there is a problem: the customer portal needs to act on behalf of users, and in order to do that, the portal needs to either have all of their private keys, or have a reliable way to ask them to sign requests. The first of these is undesirable (I hope for obvious reasons), and the second seems rather difficult with web browsers as they are today.

It would seem like a practical means to allow the customer portal to impersonate users to CloudAPI would be to use some form of shared secret between the two -- used in a form where that secret never has to be on the wire -- so an HMAC or symmetric encrypt-then-MAC scheme would be a good fit.

Keyapi seems to provide exactly such a mechanism and the code is already built into CloudAPI -- however, it is a bit legacy and has some "special" things like handling double-jsoned-json and makes a horrible hash of the normal signed Authorization header on CloudAPI's end. Nobody seems to be making good noises about it as it is, but that doesn't mean it isn't possible to clean it up and make it into something worth having in the API for system integrators developing with SDC.

So basically this issue is a request, to open up the keyapi repository as it is. It may be horrible code nobody seems to like, but it's listed as a dependency in `package.json` for sdc-cloudapi, included in the pre-built zone images, and really it should either be open, or entirely removed. My preference is strongly for it to be opened, so that anyone who's interested in helping to replace it with a better mechanism for system integrators to use can propose patches for cleaning it up.

**Comment by Marsell** [comment]
Sorry for the delay.

Okay, here's the deal: if you want to make cloudapi calls, you need to provide http-signature details signed using the caller's account (e.g. you need an account for your portal). The X-Auth-Token stuffed with the keyapi token contains the details about the user the portal is calling on behalf of.

Basically, your headers would look something like the following:

Authorization: Signature keyId="/<your portal or ops account>/keys/<some portal key>",algorithm="RSA-SHA256" <base64-encoded http-signature chunk>
X-Auth-Token: {"keyid":"<keyapi-generated UUID","data":"<base64 data about the portal's user initiating this call>","version":"0.1.0","hash":"<base64 hash of data>"}

I hope that clarifies things. My apologies that there are no docs about it.

**Comment by tgross** [comment]
@arekinath this issue has been sitting open for a while. Do we have an open JIRA ticket associated with it, or an RFD or mailing list discussion? Or is it safe to close this GitHub issue.

**Comment by arekinath** [comment]
Safe to close. The `keyapi` repo was opened up.
